pkg/store/bucket.go

```diff
@@ -417,31 +417,49 @@ type seriesEntry struct {
 type bucketSeriesSet struct {
 	set  []seriesEntry
 	i    int
+	chkI int
 	chks []storepb.Chunk
 }
 
 func newBucketSeriesSet(set []seriesEntry) *bucketSeriesSet {
 	return &bucketSeriesSet{
-		set: set,
-		i:   -1,
+		set:  set,
+		i:    -1,
+		chkI: 0,
 	}
 }
 
 func (s *bucketSeriesSet) Next() bool {
-	if s.i >= len(s.set)-1 {
+	const grpcMaxMsgSize = 4 * 1024 * 1024
+
+	if s.i >= len(s.set)-1 && s.chkI == 0 {
 		return false
 	}
-	s.i++
-	s.chks = make([]storepb.Chunk, 0, len(s.set[s.i].chks))
 
-	for _, c := range s.set[s.i].chks {
+	if s.chkI > 0 {
+		s.i++
+	}
+
+	s.chks = make([]storepb.Chunk, 0, len(s.set[s.i].chks[s.chkI:len(s.set[s.i].chks)-1]))
+
+	var bytesCount int
+	for i, c := range s.set[s.i].chks[s.chkI : len(s.set[s.i].chks)-1] {
+		bytesCount += len(c.Chunk.Bytes())
+
+		if bytesCount >= grpcMaxMsgSize {
```

Comment:
also here we could be smarter than just "SEND 1000 chunks because it is almost 4MB+.. and then.. oh send the last single one because it left"